android/envoy/src/main/kotlin/org/greatfire/envoy/EnvoyTest.kt

```diff
@@ -0,0 +1,214 @@
+//
+// The url should be the one provided by the caller
+// testType lets us know how to treat that URL
+// some transports have an additional URL, e.g. a SOCKS5 URL
+// for a PT, that gets stored in proxyUrl
+// proxyiSEnvoy lets us know if proxyUrl is an Envoy proxy or SOCKS/HTTP
+
+data class EnvoyTest(
+    var testType: EnvoyServiceType,
+    var url: String,
+) {
+    companion object {
+        private const val TAG = "EnvoyTest"
+    }
+
+    // proxy URL for the service providing transport
+    // can be SOCKS5, HTTP(S), Envoy, see proxyIsEnvoy
+    var proxyUrl: String? = null
+    // - true means the proxyUrl refers to a nonstandard Envoy proxy
+    // as documented at http://github.com/greatfire/envoy/
+    // - false means it's SOCKS or HTTP
+    var proxyIsEnvoy: Boolean = false
+
+    // Is this the service we chose to use
+    var selectedService: Boolean = false
+
+    // Is the associated service (if any) running?
+    var serviceRunning: Boolean = false
+
+    // envoy:// URL related options
+    val headers = mutableListOf<Pair<String, String>>()
+    // address param creates an resolver rule
+    // stash it here in case we can support it with OkHttp
+    var address: String? = null
+    var resolverRules: String? = null
+
+    // Envoy Global settings and state
+    private val state = EnvoyState.getInstance()
+
+    // used to time how long it takes to connect and test
+    private var timer: Timer? = null
+    // should this be in settings?
+    // private var shadowsocks: EnvoyShadowsocks? = null
+    private val shadowsocksIntent: Intent? = null
+
+    override fun toString(): String {
+        return UrlUtil.sanitizeUrl(url) + " (" + testType + ")"
+    }
+
+    private fun getTimer(): Timer {
+        if (timer == null) {
+            timer = Timer()
+        }
+        return timer!!
+    }
+
+    // helper to time things
+    inner class Timer() {
+        private val startTime = System.currentTimeMillis()
+        private var stopTime: Long? = null
+
+        fun stop(): Long {
+            stopTime = System.currentTimeMillis()
+            return stopTime!! - startTime
+        }
+
+        fun timeSpent(): Long {
+            if (stopTime == null) {
+                Log.e(TAG, "timeSpent called before stop()!")
+                return 0
+            }
+            return stopTime!! - startTime
+        }
+    }
+
+    private fun getV2RayUuid(url: String): String {
+        val san = UrlQuerySanitizer()
+        san.setAllowUnregisteredParamaters(true)
+        san.parseUrl(url)
+        return san.getValue("id")
+    }
+
+    // returns a string $host:$port where the running service can be found
+    suspend fun startService(): String {
+        if (serviceRunning) {
+            Log.e(TAG, "Tried to start $testType when it was already running")
+            return ""
+        }
+
+        Log.d(TAG, "starting service for $testType")
+
+        serviceRunning = true
```

Comment:
should this use one of those atomic classes?

Reply:
I don't think it would hurt anything if it was, but we'd likely have other problems if we're trying to start the same service twice at the same time, at least for many of the services